File: src/best_time_to_buy_and_sell_stock.rs
struct Solution;

impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let mut profit = 0;
        let mut buy = prices[0];

        for i in 1..prices.len() {
            profit = std::cmp::max(profit, prices[i] - buy);
            buy = std::cmp::min(buy, prices[i]);
        }
        profit
    }
}

fn main() {
    let prices = vec![2, 1, 2, 0, 1];
    let result = Solution::max_profit(prices);
    println!("The result is {}", result);
}

File: src/cpp/best_time_to_buy_and_sell_stock.cpp
#include <vector>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit = 0;
        int buy = prices[0];
        for (int i = 1; i < prices.size(); i++) {
            profit = std::max(profit, prices[i] - buy);
            buy = std::min(buy, prices[i]);
        }
        return profit;
    }
};

// https://leetcode.com/problems/best-time-to-buy-and-sell-stock/description/